### PR TITLE
Allow user to overwrite unknown shapes loaded from ONNX model

### LINKED ISSUE #3932: cannot use fasterrcnn onnx model in c# ap
steps:
download https://github.com/onnx/models/tree/master/faster_rcnn
try to run model in .net
```
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.ML;
using AlexNet.DataStructures;
using Microsoft.ML.Transforms.Image;

namespace SomeNet
{
    public class ImageClassifier
    {
        private readonly string dataLocation;
        private readonly string labelsLocation;
        private readonly string imagesLocation;
        private readonly string modelLocation;
        private readonly MLContext mLContext;

        public ImageClassifier(string dataLocation, string imagesLocation, string labelsLocation, string modelLocation)
        {
            this.dataLocation = dataLocation;
            this.labelsLocation = labelsLocation;
            this.imagesLocation = imagesLocation;
            this.modelLocation = modelLocation;
            mLContext = new MLContext();
        }

        public struct ImageSettings
        {
            public const int imageHeight = 224;
            public const int imageWidth = 224;
        }

        public struct OnnxModelSettings
        {
            public const string ModelInput = "image";
        }

        public void Score()
        {
            var model = LoadModel(dataLocation, imagesLocation, modelLocation);

            var predictions = PredictData(dataLocation, imagesLocation, labelsLocation, model).ToArray(); 
        }

        private PredictionEngine<ImageData, ImagePrediction> LoadModel(string dataLocation, string imagesLocation, string modelLocation)
        {
            var data = mLContext.Data.LoadFromTextFile<ImageData>(dataLocation, hasHeader: false);

            var pipeline = mLContext.Transforms.LoadImages(outputColumnName: "loadedimage", imageFolder: imagesLocation, inputColumnName: nameof(ImageData.ImagePath))
                .Append(mLContext.Transforms.ResizeImages(outputColumnName: "resizedimage",inputColumnName:"loadedimage", imageWidth: ImageSettings.imageWidth, imageHeight: ImageSettings.imageHeight))
                .Append(mLContext.Transforms.ExtractPixels(outputColumnName: "image", inputColumnName:"resizedimage", orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ABGR))
                .Append(mLContext.Transforms.ApplyOnnxModel(modelFile: modelLocation, outputColumnNames: new[] { "6379", "6381", "6383" }, inputColumnNames: new[] { "image"}));
            var model = pipeline.Fit(data);

            var predictionEngine = mLContext.Model.CreatePredictionEngine<ImageData, ImagePrediction>(model, ignoreMissingColumns:false);

            return predictionEngine;
        }

        protected IEnumerable<ImageData> PredictData(string imageLocation, string imageFolder, string labelsLocation, PredictionEngine<ImageData, ImagePrediction> model)
        {
            var labels = labelsLocation;
            var p = model.Predict(new ImageData{
                ImagePath = "/home/snake/aspose/anton.perhunov/AlexNet/AlexNet/assets/Data/html.jpg"
            });
            var im =  new ImageData{
                ImagePath = imageFolder + "/e5725c32f8be4da8b6a27375fc7669ae_html.jpg"
            };
            yield return im;
        }
    }
}
```

when inference any image you will get
Unhandled Exception: System.ArgumentException: Length of memory (150528) must match product of dimensions (3).
at System.Numerics.Tensors.DenseTensor1..ctor(Memory1 memory, ReadOnlySpan1 dimensions, Boolean reverseStride) at Microsoft.ML.Transforms.Onnx.OnnxUtils.CreateNamedOnnxValue[T](String name, ReadOnlySpan1 data, List1 shape) at Microsoft.ML.Transforms.Onnx.OnnxTransformer.Mapper.NamedOnnxValueGetterVec1.GetNamedOnnxValue()
at Microsoft.ML.Transforms.Onnx.OnnxTransformer.Mapper.UpdateCacheIfNeeded(Int64 position, INamedOnnxValueGetter[] srcNamedOnnxValueGetters, String[] activeOutputColNames, OutputCache outputCache)
at Microsoft.ML.Transforms.Onnx.OnnxTransformer.Mapper.<>c__DisplayClass13_01.<MakeGetter>b__0(VBuffer1& dst)
at Microsoft.ML.Data.TypedCursorable1.TypedRowBase.<>c__DisplayClass8_01.b__0(TRow row)
at Microsoft.ML.Data.TypedCursorable1.TypedRowBase.FillValues(TRow row) at Microsoft.ML.PredictionEngineBase2.Predict(TSrc example)

it seems that .net api does not support dynamic input sizes(-1 in dimension), it just treat them as 1, model supports inputs with (3, -1,-1), however onnx value size become (3,1,1).

### PR DESCRIPTION
**Pull request by wschin**
This PR adds an extra argument to `ApplyOnnxModel` functions.
```c#
        /// <param name="shapeDictionary">ONNX shape should be used to over those loaded from <paramref name="modelFile"/>.</param>
```
If a key-value pair, (`tensor_name`, `[1, 2, 3]`) exists, the `tensor_name`'s shape loaded from ONNX model file will be replaced by `[1, 2, 3]`. It's allows user to replace unknown shapes in their applications.

Fix #3932.

With this PR, FasterRCNN mentioned in #3932 can be run as the example below.
```c#
        private class RcnnInput
        {
            [ColumnName("image")]
            [VectorType(3, 224, 224)]
            public float[] Input { get; set; }
        }

        private class RcnnOutput
        {
            [ColumnName("6379")]
            public float[] Boxes { get; set; }

            [ColumnName("6381")]
            public long[] Label { get; set; } 

            [ColumnName("6383")]
            public float[] Score { get; set; }
        }

        static public IEnumerable<float> GenerateRandomVector(int size)
        {
            var rnd = new Random(0);

            for (int i = 0; i < size; ++i)
                yield return rnd.NextSingle();
        }

        /// <summary>
        /// A test to check if dynamic shape works.
        /// </summary>
        [OnnxFact]
        public void TestOnnxTransformDynamicShape()
        {
            var modelFile = Path.Combine(Directory.GetCurrentDirectory(), "fasterRCNN", "faster_rcnn_R_50_FPN_1x.onnx");

            var dataPoints = new RcnnInput[] {
                new RcnnInput() { Input = GenerateRandomVector(3 * 224 * 224).ToArray(), }
            };

            var shapeDictionary = new Dictionary<string, int[]>() { { "image", new int[] { 3, 224, 224 } } };

            var dataView = ML.Data.LoadFromEnumerable(dataPoints);
            var transformedDataView = ML.Transforms.ApplyOnnxModel(new[] { "6379", "6381", "6383" }, new[] { "image" }, modelFile,
                shapeDictionary: shapeDictionary).Fit(dataView).Transform(dataView);

            // Convert IDataView to IEnumerable<ZipMapOutput> and then inspect the values.
            var transformedDataPoints = ML.Data.CreateEnumerable<RcnnOutput>(transformedDataView, false).ToList();

            for (int i = 0; i < transformedDataPoints.Count; ++i)
            {
                Assert.NotNull(transformedDataPoints[i].Boxes);
                Assert.NotNull(transformedDataPoints[i].Label);
                Assert.NotNull(transformedDataPoints[i].Score);
            }
        }
```

### CODE REVIEW

File: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs [hunk 1]
```diff
@@ -226,36 +328,17 @@ private static IRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, Dat
         /// <param name="modelFile">Model file path.</param>
         /// <param name="gpuDeviceId">Optional GPU device ID to run execution on. Null for CPU.</param>
         /// <param name="fallbackToCpu">If GPU error, raise exception or fallback to CPU.</param>
-        internal OnnxTransformer(IHostEnvironment env, string modelFile, int? gpuDeviceId = null, bool fallbackToCpu = false)
```
Comment (wschin) [inline src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs]: This ctor is not used at all, so I removed it.

File: src/Microsoft.ML.OnnxTransformer/OnnxCatalog.cs [hunk 2]
```diff
@@ -33,8 +35,10 @@ public static class OnnxCatalog
         public static OnnxScoringEstimator ApplyOnnxModel(this TransformsCatalog catalog,
             string modelFile,
             int? gpuDeviceId = null,
-            bool fallbackToCpu = false)
-        => new OnnxScoringEstimator(CatalogUtils.GetEnvironment(catalog), modelFile, gpuDeviceId, fallbackToCpu);
+            bool fallbackToCpu = false,
+            IDictionary<string, int[]> shapeDictionary = null)
+        => new OnnxScoringEstimator(CatalogUtils.GetEnvironment(catalog), modelFile, gpuDeviceId, fallbackToCpu,
```
Comment (artidoro) [inline src/Microsoft.ML.OnnxTransformer/OnnxCatalog.cs]: I think this should be a breaking change. I am afraid API compat is not active for this assembly. I will open an issue about it and fix it in the meantime.
Unfortunately the signature of the function changes if you add another parameter. The only way to do this change is to add a new overload with the new parameter.

**Comment by artidoro** [comment]
>         private protected override void SaveModel(ModelSaveContext ctx)

Why are we not saving the `ShapeDictionary` here?
It seems like we will lose that information.

---
Refers to: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:307 in ced8c42. [](commit_id = ced8c4247cc35d159cad9e0a9c6649b4514d3448, deletion_comment = False)

File: test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs [hunk 3]
```diff
@@ -607,5 +635,142 @@ public void SmokeInMemoryOnnxMapTypeTest()
...
         }
+
+        /// <summary>
+        /// A test to check if dynamic shape works.
+        /// </summary>
+        [OnnxFact]
+        public void TestOnnxTransformWithCustomShapes()
+        {
+            // The loaded model has input shape [-1, 3] and output shape [-1].
+            var modelFile = Path.Combine(Directory.GetCurrentDirectory(), "unknowndimensions", "test_unknowndimensions_float.onnx");
```
Comment (artidoro) [inline test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs]: >test_unknowndimensions_float [](start = 96, length = 28)

Could we use CSharp style for file names?

Comment (artidoro) [inline test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs]: >test_unknowndimensions_float [](start = 96, length = 28)

Could you add a link to the description of the onnx model this test is using?

File: test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs [hunk 4]
```diff
@@ -607,5 +635,142 @@ public void SmokeInMemoryOnnxMapTypeTest()
...
+            for(int i = 0; i < transformedDataPoints.Count; ++i)
+                Assert.Equal(transformedDataPoints[i].argmax, expectedResults[i]);
+        }
+
+        /// <summary>
+        /// This function runs a ONNX model with user-specified shapes <paramref name="shapeDictionary"/>.
+        /// </summary>
+        /// <param name="shapeDictionary">Dictionary of tensor shapes. Keys are tensor names
+        /// while values the associated shapes.</param>
+        private void TryModwlWithCustomShapesHelper(IDictionary<string, int[]> shapeDictionary)
```
Comment (artidoro) [inline test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs]: >TryModwlWithCustomShapesHelper [](start = 21, length = 30)

I think there might be a small typo: Modwl -> Model

File: src/Microsoft.ML.OnnxTransformer/OnnxCatalog.cs (hunk 2 above)
Comment (artidoro) [reply src/Microsoft.ML.OnnxTransformer/OnnxCatalog.cs]: There is already a PR on reactivating API compat, could you approve it and rebase afterwards?

---
In reply to: [301718819](https://github.com/dotnet/machinelearning/pull/3963#discussion_r301718819) [](ancestors = 301718819)

Comment (wschin) [reply src/Microsoft.ML.OnnxTransformer/OnnxCatalog.cs]: Sure. Thanks.

---
In reply to: [301768225](https://github.com/dotnet/machinelearning/pull/3963#discussion_r301768225) [](ancestors = 301768225,301718819)

File: test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs (hunk 3 above)
Comment (wschin) [reply test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs]: It's in another repo and I feel it looks just fine.

---
In reply to: [301724967](https://github.com/dotnet/machinelearning/pull/3963#discussion_r301724967) [](ancestors = 301724967)

Comment (wschin) [reply test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs]: There is no such a description for those internal ONNX models but I can add a link to its source.

---
In reply to: [301726035](https://github.com/dotnet/machinelearning/pull/3963#discussion_r301726035) [](ancestors = 301726035)

**Comment by wschin** [comment]
>         private protected override void SaveModel(ModelSaveContext ctx)

Fixed and added a test.

---
In reply to: [509748317](https://github.com/dotnet/machinelearning/pull/3963#issuecomment-509748317) [](ancestors = 509748317)

---
Refers to: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:307 in ced8c42. [](commit_id = ced8c4247cc35d159cad9e0a9c6649b4514d3448, deletion_comment = False)

**Comment by artidoro** [comment]
>                 verWrittenCur: 0x00010002,

Should we change the version number since we are changing the binary format of how it is saved?

---
Refers to: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:155 in ab6424c. [](commit_id = ab6424c5a6a01877cc10737324e6b251552687df, deletion_comment = False)

**Comment by artidoro** [comment]
>         {

It would be very nice to have in a comment the layout of the binary format like we do in other transformers.
It's not a blocking issue though I think it would be very nice to have :)
If you do add it, you could write it in the load method above too.

---
Refers to: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:327 in ab6424c. [](commit_id = ab6424c5a6a01877cc10737324e6b251552687df, deletion_comment = False)

**Pull request merged**